keiko-sql/src/main/kotlin/com/netflix/spinnaker/q/sql/SqlDeadMessageHandler.kt

```diff
@@ -0,0 +1,82 @@
+package com.netflix.spinnaker.q.sql
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.databind.SerializationFeature
+import com.google.common.hash.Hashing
+import com.netflix.spinnaker.q.DeadMessageCallback
+import com.netflix.spinnaker.q.Message
+import com.netflix.spinnaker.q.Queue
+import de.huxhorn.sulky.ulid.ULID
+import org.jooq.DSLContext
+import org.jooq.impl.DSL
+import org.slf4j.LoggerFactory
+import java.lang.Exception
+import java.nio.charset.StandardCharsets
+import java.time.Clock
+
+class SqlDeadMessageHandler(
+  deadLetterQueueName: String,
+  schemaVersion: Int,
+  private val jooq: DSLContext,
+  private val clock: Clock
+) : DeadMessageCallback {
+
+  companion object {
+    @Suppress("UnstableApiUsage")
+    private val hashObjectMapper = ObjectMapper().copy().apply {
+      enable(SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS)
+    }
+
+    private val nameSanitization = """[^A-Za-z0-9_]""".toRegex()
+
+    private val log = LoggerFactory.getLogger(SqlDeadMessageHandler::class.java)
+  }
+
+  private val dlqBase = "keiko_v${schemaVersion}_dlq"
+  private val dlqTableName = "${dlqBase}_${deadLetterQueueName.replace(nameSanitization, "_")}"
+
+  private val dlqTable = DSL.table(dlqTableName)
+  private val idField = DSL.field("id")
+  private val fingerprintField = DSL.field("fingerprint")
+  private val updatedAtField = DSL.field("updated_at")
+  private val bodyField = DSL.field("body")
+
+  init {
+    initTables()
+  }
+
+  override fun invoke(queue: Queue, message: Message) {
+    var fingerprint: String? = null
+    var json: String? = null
+
+    try {
+      /* Storing the fingerprint may be useful for correlating Queue log messages to DLQ rows */
+      fingerprint = message.fingerprint()
+      json = hashObjectMapper.writeValueAsString(message)
+
+      jooq.insertInto(dlqTable)
+        .set(idField, ULID().nextValue().toString())
```

Comment:
Should promote `ULID` into a class property so we don't take the initialization cost every invocation.